### PR TITLE
core/state, eth/downloader, trie: reset fast-failure on progress

### PR DESCRIPTION
**Pull request by karalabe**
The recent state bloat surfaced a tech debt that the downloader and/or fast sync had.

Originally if fast sync aborted in the critical section, geth considered it an attack and reverted to full sync. This turned out to cause issues for many people with slower internet connections as there were occasional hiccups in the state downloads that we did at the end of the chain, causing spurious revertals.

A later change fixed this issue by starting state trie downloading concurrently with header/body/receipt downloads. This caused the state trie to be fully downloaded by the time we reached the critical section. For cases where there were a few nodes still pending, we introduced a retry counter that allowed up to 10 sync failures. This solved the issue for people.

The recent state bloat however caused fast sync to be unable to **ever** download the entire state trie before the critical section, leaving a large part of it to be pulled during that sensitive period. Although we did have 10 trials to restart, downloading 30M entries will abort usually more than 10 times until it's fully pulled, so fast sync started to revert much much more often.

This PR fixes the issue by bumping the fail allowance to 32 (from 10) and also starting to track sync progression: if a sync cycle managed to write even 1 state entry into the database, it means that fast sync does indeed do something useful, so the failure counter is reset. This should ensure that no matter how slow, as long as you can progress, fast sync should never revert to slow sync.

---

The second commit reduces the fast sync block retention from 1K to 64, and the random pivot selection interval from 512 to 256. This should help fast sync be less dependent on the generosity of other client implementations which don't play too nice. **Per the calculations in https://github.com/ethereum/go-ethereum/issues/1889, the 1K full blocks had no relevance to the security of fast sync, the 24 PoW verification that must be performed are always enforced**.

The annoying downloader test that occasionally failed should hopefully be more or less fixed too.

---

The third commit fixes a user annoyance with a log that was displayed when switching from fast syncing the head to fast syncing the pivot state.

---

The fourth commit is a fix that reduces a few hundred stale network requests that happened during the switch-over from head-trie-retrieval to pivot-trie-retrieval.

Until now we only reset the `trie` sync scheduler, but did not abort already in flight and/or queued state retrievals. This resulted in retrieving a up to 8192 state entries uselessly (that were already queued before the switch-over). This commit drops those out of our internal queues.

Note that if a request is already in flight, we can't simply forget about it as that would result in an immediate new concurrent request being made, so for in-flight requests we just invalidate them, but nonetheless wait for their completion.

### CODE REVIEW

File: eth/downloader/downloader.go [hunk 1]
```diff
@@ -951,6 +956,11 @@ func (d *Downloader) fetchNodeData() error {
 				syncStatsStateDone := d.syncStatsStateDone // Thread safe copy for the log below
 				d.syncStatsLock.Unlock()
 
+				// If real database progress was made, reset any fast-sync pivot failure
+				if progressed && atomic.LoadUint32(&d.fsPivotFails) > 1 {
```
Comment (Arachnid) [inline eth/downloader/downloader.go]: I assume the point of using atomic here is to guard against race conditions of some sort, but this if/then statement itself introduces a race condition, since the read/modify/write is not atomic.

File: eth/downloader/downloader_test.go [hunk 2]
```diff
@@ -37,33 +37,87 @@ import (
...
 	testKey, _  = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
 	testAddress = crypto.PubkeyToAddress(testKey.PublicKey)
-	genesis     = core.GenesisBlockForTesting(testdb, testAddress, big.NewInt(1000000000))
 )
 
 // Reduce some of the parameters to make the tester faster.
 func init() {
 	MaxForkAncestry = uint64(10000)
 	blockCacheLimit = 1024
+	fsCriticalTrials = 10
```
Comment (Arachnid) [inline eth/downloader/downloader_test.go]: Are these test changes relevant to the fast sync improvements? If not, could they be made into a separate PR?

Comment (karalabe) [inline eth/downloader/downloader_test.go]: Yes, the fast sync changes blow up the test.

File: eth/downloader/downloader.go (hunk 1 above)
Comment (karalabe) [inline eth/downloader/downloader.go]: That's not really a problem. I just want to check whether I can reset it to 1 or not. If it gets reset twice by accident so be it, no biggie.
